ClipboardAPI/clipboard-change-event-explainer.md

```diff
@@ -0,0 +1,189 @@
+
+#### 5.2.3 Approach 3 - Transient user activation
+This approach allows the clipboardchange event to be fired for a short duration after the user loses page focus, such as up to 5 seconds. This ensures that clipboard changes occurring immediately after focus loss are still captured, enhancing user experience without compromising security.
+
+##### Pros:
+1. Clipboard changes occurring immediately after the user loses focus are still captured, ensuring the web app can respond promptly when the user returns.
+Example: A web app can pre-process clipboard data while the user is in another application, reducing wait time when the user comes back.
+2. Limits the duration for which clipboard monitoring is allowed after focus loss, reducing the risk of prolonged unauthorized access.
+
+##### Cons:
+1. The short duration might not be sufficient for some use cases where clipboard changes occur after the specified time.
+2. Still requires monitoring for a brief period after focus loss, which could lead to resource usage if many pages implement this.
+Example: Multiple tabs monitoring clipboard changes for 5 seconds could still cause a temporary spike in resource usage.
+
+#### 5.2.4 Conclusion
+We favour Approach 1 - Page required to be in focus to receive event, since this approach is inline with the current Async clipboard APIs and also reduces the possibility of misusing the clipboard change event in privacy related attacks. Also the approach has relatively lower resource usage.
+
+### 5.3 Event details 
+Since the clipboardchange event is not triggered by a user action and the event is not associated to any DOM element, hence this event doesn't bubbles and is not cancellable.
+
+There are two ways to get the changed clipboard data within the event handler:
+
+1. Async Clipboard API - The [read](https://w3c.github.io/clipboard-apis/#dom-clipboard-read) or [readText](https://w3c.github.io/clipboard-apis/#dom-clipboard-readtext) methods of the [Async clipboard API](https://w3c.github.io/clipboard-apis/#async-clipboard-api) can be used to get the current contents of the system clipboard.
+
+2. DataTransfer API - The clipboardchange event is a [ClipboardEvent](https://www.w3.org/TR/clipboard-apis/#clipboard-event-interfaces) that includes a [DataTransfer](https://html.spec.whatwg.org/multipage/dnd.html#datatransfer) object. This keeps the interface of this event consistent with other clipboard related events like [cut](https://w3c.github.io/clipboard-apis/#clipboard-event-cut), [copy](https://w3c.github.io/clipboard-apis/#clipboard-event-copy) or [paste](https://w3c.github.io/clipboard-apis/#clipboard-event-paste) events. The [getData](https://html.spec.whatwg.org/multipage/dnd.html#dom-datatransfer-getdata) method of DataTransfer interface can be used to retrieve the clipboard contents of a specific format.
```

Comment:
Hi! Regarding the option to contain the changed data in the event, I'm not sure of that's a great idea from two perspectives:
- Single responsibility: The purpose of this event is to notify *that* change(s) happened, not necessary *what* change(s). Filling this field with contents of the clipboard would duplicate the functionality of `navigator.read*` and provide yet another way to read the clipboard. Also, [clipboardData](https://w3c.github.io/clipboard-apis/#clipboard-event-interfaces:~:text=%3F%20clipboardData%3B%0A%7D%3B-,clipboardData,-The%20clipboardData%20attribute) field documentation seems to suggest that it's to be used for copy, cut and paste events (in which this could be used to modify the data before it reaches its destination).
- Efficiency and accuracy: Filling an event with clipboard contents after every change and sending such event to every site that has the clipboard permission sounds like it could be really inefficient, especially in case of a big clipboard size. And this data might not even be relevant - for example, if the app would like to explicitly react by reading the clipboard using `navigator.clipboard.read` with specified `formats`, this might be entirely useless.

Reply:
Thanks for reviewing. It does make sense to have this event not inherit the ClipboardEvent interface due to the reasons you mentioned. I will update the doc accordingly.

Reply:
I mean, for consistency with the whole Clipboard Event API I would keep this event as an instance of the ClipboardEvent class - but it might be a good idea to take advantage of the fact that this field is null by default and just not set it:
`DataTransfer? clipboardData = null;`
Alternatively, like in [this case](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/editing/commands/clipboard_commands.cc;drc=7bf0ba43d87a09d7a866a78cec04986d957747b8;l=189) - an already existing example of not filling this field - just put there an empty object.

Reply:
Agree